src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1741,27 +1750,75 @@ public override IEnumerator GetEnumerator()
-                ? new List<NpgsqlDbColumn>().AsReadOnly()
-                : new DbColumnSchemaGenerator(_connection, RowDescription, _behavior.HasFlag(CommandBehavior.KeyInfo))
-                    .GetColumnSchema();
+            => GetColumnSchema(false).GetAwaiter().GetResult();
 
 #if !NET461
         ReadOnlyCollection<DbColumn> IDbColumnSchemaGenerator.GetColumnSchema()
             => new ReadOnlyCollection<DbColumn>(GetColumnSchema().Cast<DbColumn>().ToList());
 #endif
 
+        /// <summary>
+        /// Asynchronously returns schema information for the columns in the current resultset.
+        /// </summary>
+        /// <returns></returns>
+        public Task<ReadOnlyCollection<NpgsqlDbColumn>> GetColumnSchemaAsync(CancellationToken cancellationToken = default)
+            => GetColumnSchema(true, cancellationToken).AsTask();
+
+        ValueTask<ReadOnlyCollection<NpgsqlDbColumn>> GetColumnSchema(bool async, CancellationToken cancellationToken = default)
+            => RowDescription == null || RowDescription.Fields.Count == 0
+                ? new ValueTask<ReadOnlyCollection<NpgsqlDbColumn>>(new List<NpgsqlDbColumn>().AsReadOnly())
+                : new DbColumnSchemaGenerator(_connection, RowDescription, _behavior.HasFlag(CommandBehavior.KeyInfo))
+                    .GetColumnSchemaAsync(async, cancellationToken);
+
+#if !NET461 && !NETSTANDARD2_0 && !NETSTANDARD2_1 && !NETCOREAPP3_0
+        public /*override*/ Task<ReadOnlyCollection<DbColumn>> GetColumnSchemaAsync(CancellationToken cancellationToken = default)
+#else
+        public Task<ReadOnlyCollection<DbColumn>> GetColumnSchemaAsync(CancellationToken cancellationToken = default)
+#endif
+            => GetColumnSchemaAsync(cancellationToken).ContinueWith(t => new ReadOnlyCollection<DbColumn>(t.Result.Cast<DbColumn>().ToList()));
         #endregion
 
         #region Schema metadata table
 
-
         /// <summary>
         /// Returns a System.Data.DataTable that describes the column metadata of the DataReader.
         /// </summary>
 #nullable disable
         public override DataTable GetSchemaTable()
 #nullable restore
+        {
+            var table = GetEmptySchemaTable();
+
+            foreach (var column in GetColumnSchema())
+            {
+                AddColumnToSchemaTable(table, column);
+            }
+
+            return table;
+        }
+
+        /// <summary>
+        /// Asynchronously returns a System.Data.DataTable that describes the column metadata of the DataReader.
+        /// </summary>
+#nullable disable
+#if !NET461 && !NETSTANDARD2_0 && !NETSTANDARD2_1 && !NETCOREAPP3_0
+        public /*override*/ async Task<DataTable> GetSchemaTableAsync(CancellationToken cancellationToken = default)
+#else
+        public async Task<DataTable> GetSchemaTableAsync(CancellationToken cancellationToken = default)
+#endif
+#nullable restore
+        {
+            var table = GetEmptySchemaTable();
```

Comment:
This is missing the regular async boilerplate for async methods (i.e. check for cancellation token, NoSynchronizationContextScope). See other async methods as an example.

Reply:
Good remark 😋
Done.